### PR TITLE
Integrate JS asset build into Golang Dockerfile to ensure consistent builds and deployment process"

### PR DESCRIPTION
**Pull request by shubhadapaithankar**
### Which issue this PR addresses:
Fixes [ARO-7157 ](https://issues.redhat.com/browse/ARO-7157)

### What this PR does / why we need it:

- This PR merges the JavaScript asset build process with Golang build, ensuring that all components are build in a single, streamlined process.
-  It ensures JS assets are available before running `go generate` and uses a strict NodeJS version image for the JS build.
- Merges `Dockerfile.ci-unified` , `Dockerfile.ci-portal` and `Dockerfile.ci-rp `into a single `Dockerfile.ci-rp`, integrating the JavaScript asset build process with the Golang build.
- Updates the Makefile to modify the `ci-rp `target for building with the merged `Dockerfile.ci-rp`.

### Test plan for issue:
Execute the `make ci-rp `command to build the Docker image and verify that the final RP image is correctly produced with all necessary assets and functionalities intact.

### Is there any documentation that needs to be updated for this PR?
N/A

### How do you know this will function as expected in production? 
Currently not a production change - this will be a parallel CI/CD effort that for now does not impact prod.

### CODE REVIEW

File: Dockerfile.ci-unified [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+#Copying package files and installing dependancies
+COPY portal/v2/package*.json ./
+RUN npm ci
+
+#Copying the rest of the source and build
+COPY portal/v2/ ./
+RUN npm run lint
+RUN mkdir -p /build/portal/v2/dist  # Ensure the directory exists
+RUN npm run build || (echo "Build failed, listing current directory:" && ls -la)
+
```
Comment (SudoBrendan) [inline Dockerfile.ci-unified:16]: bug: when I run this locally `make ci-rp` does not execute this stage. https://docs.docker.com/build/building/multi-stage/ is worth a read here to understand why (ignore the `buildkit` stuff - podman doesn't support a majority of it) :) and reach out if you need additional help.

File: Dockerfile.ci-unified [hunk 2]
```diff
@@ -0,0 +1,52 @@
+ARG REGISTRY
```
Comment (SudoBrendan) [inline Dockerfile.ci-unified:1]: bug: `VERSION` has been dropped - so the RP is no longer being passed a git commit.

File: Dockerfile.ci-unified
Comment (SudoBrendan) [inline Dockerfile.ci-unified:1]: question: I'd be curious if people would find value in different dockerfiles for portal, RP, and a unified build, or if we should just overwrite `Dockerfile.ci-rp`, and delete `Dockerfile.ci-portal`.

One implication of creating a new file is that `ci-unified` isn't paying attention to the `Dockerfile.ci-rp.dockerignore` file like the old `make ci-rp` was, so our caching context gets invalidated more often.

**Review by SudoBrendan — CHANGES_REQUESTED** [review]
Thanks Shubhada for your contribution, this is a great start and headed in the right direction! I think there's a few bugs to fix here - reach out if you want to pair on these or need additional help.

File: Dockerfile.ci-unified (hunk 2 above)
Comment (shubhadapaithankar) [reply Dockerfile.ci-unified:1]: fix this issue

File: Dockerfile.ci-unified (hunk 1 above)
Comment (shubhadapaithankar) [reply Dockerfile.ci-unified:16]: Fix this issue in my latest commit

File: Dockerfile.ci-unified
Comment (mociarain) [reply Dockerfile.ci-unified:1]: I think we should split them out. AFAIK we wire them together like this to enable the current VMSS setup and as I understand it, a major goal for us is to have nice independent images running as nice independent services in a kubernetes context in the future.

With this goal in mind a set of dockerfiles with a very thin unification step would be ideal IMO. (I think this runs counter to the idea of this ticket so I'm probably missing some context but there you go :D)

Comment (shubhadapaithankar) [reply Dockerfile.ci-unified:1]: I merged Dockerfile.ci-unified and Dockerfile.ci-rp into a single Dockerfile.ci-rp, deleting Dockerfile.ci-portal. This simplifies the build process by ensuring JS assets are built before the Golang build.

File: Dockerfile.ci-rp.dockerignore [hunk 3]
```diff
@@ -10,4 +10,17 @@
...
+!portal/v2/*.js
+!portal/v2/*.jsx
+!portal/v2/*.ts
+!portal/v2/*.tsx
+!portal/v2/*.css
+!portal/v2/*.scss
+!portal/v2/*.html
+!portal/v2/*.json
+!portal/v2/public/
+!portal/v2/src/
```
[the remark this answers is not in the document]
Comment (shubhadapaithankar) [reply Dockerfile.ci-rp.dockerignore]: I have updated the` .dockerignore `file to explicitly exclude the node_modules folder to ensure it is not copied into the container.

File: Dockerfile.ci-rp [hunk 4]
```diff
@@ -2,20 +2,50 @@ ARG REGISTRY
 ARG VERSION
 
 ###############################################################################
+# Stage 1: Build the SRE Portal Assets
 # builder is responsible for all compilation and validation of the RP
 ###############################################################################
-FROM ${REGISTRY}/ubi8/go-toolset:1.20.12-5 as builder
+FROM ${REGISTRY}/ubi8/go-toolset:1.20.12-5 as portal-build
```
[the remark this answers is not in the document]
Comment (shubhadapaithankar) [reply Dockerfile.ci-rp]: Thanks for pointing me. I have updated the base image to` /ubi8/nodejs-16` as we are not performing any Go build tasks or operations within this stage.

**Review by tsatam — DISMISSED** [review]

**Review by slawande2 — DISMISSED** [review]

File: Dockerfile.ci-rp [hunk 5]
```diff
@@ -25,24 +55,26 @@ COPY cmd cmd
 COPY pkg pkg
 COPY test test
 
-# lint
-RUN golangci-lint run --verbose
+# Ensure JS assets are available before generating Go code
+COPY --from=portal-build /build/portal/v2/${BUILD_PATH} /app/pkg/portal/assets/v2/build
```
Comment (SudoBrendan) [inline Dockerfile.ci-rp:59]: bug: ${BUILD_PATH} is not definined at this point of the multi-stage build. I think we should just hard-code the path.

**Review by SudoBrendan — CHANGES_REQUESTED** [review]

File: Dockerfile.ci-rp [hunk 6]
```diff
@@ -25,24 +55,26 @@ COPY cmd cmd
...
 RUN hack/fips/validate-fips.sh ./aro
 
 ###############################################################################
-# final is our slim image with minimal layers and tools
+# Stage 3: final is our slim image with minimal layers and tools
 ###############################################################################
 FROM ${REGISTRY}/ubi8/ubi-minimal AS final
 RUN microdnf update && microdnf clean all
 COPY --from=builder /app/aro /app/e2e.test /usr/local/bin/
+COPY --from=builder /app/cmd/aro /app/cmd/aro
```
[the remark this answers is not in the document]
Comment (shubhadapaithankar) [reply Dockerfile.ci-rp]: Removed the redundant `COPY

File: Dockerfile.ci-rp [hunk 7]
```diff
@@ -2,20 +2,50 @@ ARG REGISTRY
...
+COPY portal/v2/package*.json ./
+RUN npm ci
+RUN npm audit --omit=dev # Run audit without dev dependencies
+
+# Copying the rest of the source and build
+COPY portal/v2/ ./
+RUN npm run lint
+
+# Set permissions to avoid EACCES error
+RUN chown -R root:root /build/portal/v2
```
[the remark this answers is not in the document]
Comment (shubhadapaithankar) [reply Dockerfile.ci-rp]: Consolidated the permissions setting with the COPY --chown command to simplify and streamline the Dockerfile.

File: Dockerfile.ci-rp [hunk 8]
```diff
@@ -2,20 +2,50 @@ ARG REGISTRY
...
+
+# Set the build path
+ENV BUILD_PATH=../../pkg/portal/assets/v2/build
```
[the remark this answers is not in the document]
Comment (shubhadapaithankar) [reply Dockerfile.ci-rp]: Thanks for pointing me , The path has been removed

File: Dockerfile.ci-rp [hunk 9]
```diff
@@ -2,20 +2,50 @@ ARG REGISTRY
...
 
-# golang config
+# golang config and build steps
 ENV GOPATH=/root/go
 ENV GOFLAGS="-tags=containers_image_openpgp,exclude_graphdriver_btrfs,exclude_graphdriver_devicemapper"
 
-# install general build tools
+# Install golangci-lint and verify
 RUN curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b /usr/local/bin v1.56.2
+RUN which golangci-lint || (echo "golangci-lint not found" && exit 1)
```
[the remark this answers is not in the document]
Comment (shubhadapaithankar) [reply Dockerfile.ci-rp]: Removed the redundant lines to minimize image layers as the version specification above is sufficient.

File: Dockerfile.ci-rp.dockerignore [hunk 10]
```diff
@@ -10,4 +10,16 @@
...
+# Allow files and directories for JavaScript build
+!portal/v2/package*.json
+!portal/v2/*.js
+!portal/v2/*.jsx
+!portal/v2/*.ts
+!portal/v2/*.tsx
+!portal/v2/*.css
+!portal/v2/*.scss
+!portal/v2/*.html
+portal/v2/node_modules/
```
[the remark this answers is not in the document]
Comment (shubhadapaithankar) [reply Dockerfile.ci-rp.dockerignore]: Updated the file

File: Dockerfile.ci-rp (hunk 5 above)
Comment (shubhadapaithankar) [reply Dockerfile.ci-rp]: Fixed the bug by hardcoding the path in the COPY command to ensure the JS assets are available before generating the Go code.

**Review by slawande2 — APPROVED** [review]

**Review by anshulvermapatel — DISMISSED** [review]

**Review by rajdeepc2792 — DISMISSED** [review]
The build works correctly for the MacOS. LGTM.

File: Dockerfile.ci-rp [hunk 11]
```diff
@@ -25,24 +44,25 @@ COPY cmd cmd
...
 RUN hack/fips/validate-fips.sh ./aro
 
 ###############################################################################
-# final is our slim image with minimal layers and tools
+# Stage 3: final is our slim image with minimal layers and tools
 ###############################################################################
 FROM ${REGISTRY}/ubi8/ubi-minimal AS final
 RUN microdnf update && microdnf clean all
-COPY --from=builder /app/aro /app/e2e.test /usr/local/bin/
+COPY --from=builder /app/aro /app/e2e.test /usr/local/bin/
```
Comment (tsatam) [inline Dockerfile.ci-rp:65]: looks like some extra whitespace got included here?

File: Dockerfile.ci-rp [hunk 12]
```diff
@@ -2,20 +2,39 @@ ARG REGISTRY
...
+
+# Copying package files and installing dependencies
+COPY portal/v2/package*.json ./
+RUN npm ci
+RUN npm audit --omit=dev # Run audit without dev dependencies
+
+# Copying the rest of the source and build
+COPY --chown=root:root portal/v2/ ./
+RUN npm run lint && \
+    npm run build || (echo "Build failed, listing current directory:" && ls -la)
```
Comment (tsatam) [inline Dockerfile.ci-rp:20]: Is the `ls` output useful here? IMO we should just let the build errors stand on their own. 

```suggestion
RUN npm run lint && npm run build
```

**Review by tsatam — CHANGES_REQUESTED** [review]

File: Dockerfile.ci-rp (hunk 11 above)
Comment (shubhadapaithankar) [reply Dockerfile.ci-rp]: Fix it

**Review by rajdeepc2792 — APPROVED** [review]

**Review by tsatam — DISMISSED** [review]

**Review by cadenmarchese — DISMISSED** [review]
builds fine for me. lgtm!

**Review by slawande2 — DISMISSED** [review]

File: Dockerfile.ci-rp [hunk 13]
```diff
@@ -25,20 +43,21 @@ COPY cmd cmd
 COPY pkg pkg
 COPY test test
 
-# lint
-RUN golangci-lint run --verbose
+# Ensure JS assets are available before generating Go code
+COPY --from=portal-build /build/pkg/portal/assets/v2/ /app/pkg/portal/assets/v2/build
```
Comment (SudoBrendan) [inline Dockerfile.ci-rp:47]: bug: still have an invalid path here, this ends up creating `/app/pkg/portal/assets/v2/build/build` directory in the rp-builder image stage.

I think we want

```
COPY --from=portal-build /build/pkg/portal/assets/v2/build /app/pkg/portal/assets/v2/build
```

**Review by SudoBrendan — CHANGES_REQUESTED** [review]

File: Dockerfile.ci-rp (hunk 13 above)
Comment (shubhadapaithankar) [reply Dockerfile.ci-rp]: Fix the path

**Review by slawande2 — DISMISSED** [review]

**Comment by SudoBrendan** [comment]
Perfection. I know this is 100% correct, because I can do this:

```
# remove the currently committed portal build tracked in git
rm -rf ./pkg/portal/assets/v2/build/

# build the portal using containers
make ci-rp

# copy the build from the containerized build intermediary image layer back into my local
podman unshare
mnt=$(podman image mount f3485711b92e)
cp -R ${mnt}/app/pkg/portal/assets/v2/build ./pkg/portal/assets/v2/build

# ...and there's no git diff!
git diff
```

Excellent. LGTM!

**Review by SudoBrendan — APPROVED** [review]

File: Dockerfile.ci-rp.dockerignore [hunk 14]
```diff
@@ -10,4 +10,14 @@
 !pkg
 !swagger
 !test
-!vendor
\ No newline at end of file
+!vendor
+
```
Comment (SudoBrendan) [inline Dockerfile.ci-rp.dockerignore:14]:
```suggestion
# Don't pay attention to locally built NPM artifacts - we'll bring our own
+pkg/portal/assets/v2/build
```

...I take it back. I'm so convinced by your improvements that I say we should safely _ignore_ the old way of doing things. This makes a statement: with the new process - we have zero dependency on the old NPM build, and it works the same! This is HUGE! :)

**Review by SudoBrendan — CHANGES_REQUESTED** [review]
I'm so convinced by your build, that I think we should ignore the old way of building portal in the new one. For now, we can't `.gitignore` the portal assets... but we can at least `.dockerignore` them :muscle: My validation above still works with the `.dockerignore` changes.

File: Dockerfile.ci-rp.dockerignore (hunk 14 above)
Comment (shubhadapaithankar) [reply Dockerfile.ci-rp.dockerignore:14]: Thanks for pointing, Update Dockerfile.ci-rp.dockerignore to ignore old NPM artifacts

**Review by slawande2 — APPROVED** [review]

**Review by rajdeepc2792 — APPROVED** [review]
The build works fine at my end(MacOS). LGTM.

**Pull request merged**